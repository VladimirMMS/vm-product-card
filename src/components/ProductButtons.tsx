import React from "react";
import { useCallback, useContext } from "react";
import { productContext } from "../context/ProductContext";
import styles from "../styles/styles.module.css"

export interface ButtonProps {
  className?: string;
  style?: React.CSSProperties
}

export const ProductButtons = ({ className, style }: ButtonProps) => {
  const context = useContext(productContext);
  const { count, increaseBy, maxCount } = context;
  const isMaxReached = useCallback(
    () => { 
      if(count === maxCount) {
        return true
      }
      return false
    },
    [count, maxCount],
  )
  
  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        -
      </button>
      <div className={styles.countLabel}>{count}</div>
      <button className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`} onClick={() => increaseBy(1)}>
        +
      </button>
    </div>
  );
};
