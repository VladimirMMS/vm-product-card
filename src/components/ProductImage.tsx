import { useContext } from "react";
import styles from "../styles/styles.module.css";
import noImage from '../assets/no-image.jpg'
import { productContext } from "../context/ProductContext";
import React from "react";


export interface Props {
  img?: string;
  className?: string;
}

export const ProductImage = ({ img = '', className }: Props) => {
  const context = useContext(productContext);
  const { product } = context;
  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={img ? img : product.img ? product.img : noImage}
      alt="Product"
    />
  )
}

