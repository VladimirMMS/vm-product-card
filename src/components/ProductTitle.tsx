import React from "react";
import { useContext } from "react";
import { productContext } from "../context/ProductContext";
import styles from "../styles/styles.module.css";

export interface ProductTitleProps {
  title?: string;
  className?: string;
}

export const ProductTitle = ({ title, className }: ProductTitleProps) => {
  const context = useContext(productContext);
  const { product } = context;
  return (
    <span className={`${styles.productDescription} ${className}`}>
      {title ? title : product.title}
    </span>
  );
};
