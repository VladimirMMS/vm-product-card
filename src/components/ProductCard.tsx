import styles from "../styles/styles.module.css";
import { useProduct } from "../hooks/useProduct";
import { InitialValues, onChangeArgs, Product, ProductCartHandlers } from "../interfaces";
import { productContext } from "../context/ProductContext";
import React from "react";

export interface Props {
  product: Product;
  children: (args: ProductCartHandlers) => JSX.Element
  className?: string
  style?: React.CSSProperties
  onChange?: (args: onChangeArgs) => void
  value?: number;
  initialValues?: InitialValues
}

export function ProductCard({ children, product, className, style, onChange, value, initialValues }: Props) {
  const { count, increaseBy, maxCount, reset , isMaxCountReaced} = useProduct({ onChange, product, value, initialValues })
  const { Provider } = productContext;
  return (
    <Provider value={{
      count,
      increaseBy,
      product,
      maxCount
    }
    }>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count,
          isMaxCountReaced,
          maxCount,
          product,
          increaseBy,
          reset

        })}
      </div>
    </Provider>
  );
}

