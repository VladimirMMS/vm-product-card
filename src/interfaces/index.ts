import { ButtonProps } from "../components/ProductButtons";
import { Props as ProductCardProps } from "../components/ProductCard";
import { Props as ImageProps } from "../components/ProductImage";
import { ProductTitleProps } from "../components/ProductTitle";

export interface Product {
  id: string;
  title: string;
  img?: string;
}

export interface ProductContextProps {
  count: number;
  increaseBy: (value: number) => void;
  product: Product;
  maxCount?: number
}

export interface ProductCardHOCProps {
  ({ children, product }: ProductCardProps): JSX.Element;
  Title: ({ title, className }: ProductTitleProps) => JSX.Element;
  Image: ({ img, className }: ImageProps) => JSX.Element;
  Buttons: ({ className }: ButtonProps) => JSX.Element;
}


export interface onChangeArgs {
  product: Product,
  count: number
}

export interface ProductInCart extends Product {
  count: number;
}

export interface InitialValues {
  count?: number
  maxCount?: number
}

export interface ProductCartHandlers {
  count: number;
  isMaxCountReaced?: boolean;
  maxCount?: number;
  product: Product;
  increaseBy: (value: number) => void;
  reset: () => void
}