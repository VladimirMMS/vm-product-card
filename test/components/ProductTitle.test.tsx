import React from 'react'
import renderer from 'react-test-renderer'
import {ProductTitle, ProductCard} from '../../src/components'
import { product1 } from '../data/product';

describe('ProductTitle', () => {
    test('It should  show the component correctly with the title customed', () => {
        const wrapper = renderer.create(
            <ProductTitle title='Custom' className='custom-style'/>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();

    })
    test('It should show the product component name', () => {
        const wrapper = renderer.create(
            <ProductCard product={product1}>
                {
                    () => (
                        <ProductTitle title='Custom'/>
                    )
                }

            </ProductCard>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();


    })
  });