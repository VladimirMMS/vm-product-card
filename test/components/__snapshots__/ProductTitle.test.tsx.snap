// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle It should  show the component correctly with the title customed 1`] = `
<span
  className="productDescription custom-style"
>
  Custom
</span>
`;

exports[`ProductTitle It should show the product component name 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Custom
  </span>
</div>
`;
